### ISSUE TITLE
[9] SimCLR : A Simple Framework for Contrastive Learning of Visual Representations

### ISSUE THREAD

**Issue by long8v** [long8v/PTIR#9]
![image](https://user-images.githubusercontent.com/46675408/150898229-b10a1488-0bca-4385-8fda-81142625d61b.png)
[arxiv](https://arxiv.org/pdf/2002.05709.pdf), [code](https://github.com/google-research/simclr)
**problem :** supervised learning을 하기 위해 큰 annotation 비용. 이전의 SSL은 복잡한 아키텍쳐를 가지고 있음.
**solution :** 이미지를 augmentation하여 원래 이미지와 augmented image가 같은 이미지로 학습되도록 학습. 이미지에서 representation을 하고 이를 non-linear로 transform한 뒤 두 pair의 내적 곱의 log softmax가 최대화(=크로스엔트로피 최소화)하도록 하는 contrastive loss 적용. 
**result :** self-supervision model의 representation에 linear를 연결하여 분류하는 태스크에서 SOTA, 실제 데이터의 1%를 사용하여 fine-tuning한 모델이 top-5에서 AlexNet보다 더 좋은 성능
**details :**
![image](https://user-images.githubusercontent.com/46675408/150899388-d2f16eb6-c653-4698-90bd-9ce01616aef0.png)
. f( ) : representation, 논문에서 ResNet 사용
. g( ) : projection head, 논문에서 MLP + ReLU사용, finetuning시에는 사용하지 않음 
. contrastive loss는 x_i가 주어졌을 때, positive pair인 x_j를 구하도록 함
![image](https://user-images.githubusercontent.com/46675408/150900206-812c71de-9434-41ed-995d-eeb52aeb9b38.png)
. 이때 similiarity 함수는 v와 u의 내적

- SSL에서는 더 강한 augmentation이 필요했고, 두개 이상의 augmentation을 적용해야 했고, random crop + color distiortion 조합이 가장 성능이 좋았음
![image](https://user-images.githubusercontent.com/46675408/150901300-4d82d9dd-6a1f-4129-acbf-1c8ea31e5d9b.png)

- representation에 non-linear transformation을 추가하는 것이 성능에 크게 기여함(linear 보다 3%, 없는것보다 10%이상)
- 또한 projection을 하기 전의 h를 사용하는 것이 g(z)를 사용하는 것보다 성능이 10%이상 차이남 -> representation이 잘됐다는 의미
- contrastive loss이 성능에 크게 기여함
- supervised learning보다 더 큰 batch size(bs:8192까지), 더 긴 학습 시간이 필요했음
- LARS optimizer 사용
- 한 device에서 positive pair가 들어가도록 디자인되기 때문에 로컬 정보가 분류 모델에 적용될 수 있음. 이를 방지하기 위해 모든 device의 mean과 std를 구하는 global batch normalization이 사용됨
-

**Comment by long8v** [comment]
# [code](https://github.com/AndrewAtanov/simclr-pytorch)
official 코드는 [code](https://github.com/google-research/simclr)지만 tensorflow로 구현되어 있어, 학습을 위해 위의 코드로 봄

## Readme
- LARS를 쓰지 않고 다른 optimizer를 사용하면 weight decay를 바꿔야한다.
- 논문의 implementation의 batch size가 매우 큰데, 작게해도 temperature, lr, projection head depth등을 잘 설정하면 학습이 된다

## models
### models/resnet.py
torchvision에서 `models.resnet.ResNet` 모델을 상속함. (self.layer1, ... 는 부모 클래스에서 나온 것)
argument는 아래와 같음
- block
  - ResNet18은 `models.resnet.BasicBlock`사용, 3x3 conv -> BN2D -> ReLU -> 3x3 conv -> BN2D (-> Downsample) -> ReLU
  - ResNet50은 `models.resnet.Bottleneck`사용, 1x1 conv -> BN2D -> ReLU -> 3x3 conv -> BN2D (-> Downsample) -> ReLU
  - https://pytorch.org/vision/0.8/_modules/torchvision/models/resnet.html
- cifar_head
  - cifar_head가 True면 self.conv1 = nn.Conv2d(3, 64, kernel_size=3, stride=1, padding=1, bias=False)로 바꾸고 
  - cifar_headr가 False면 maxpooling함

flatten이라는 nn.Module을 정의하여 마지막 output에 대해 dim=1부터 [flatten](https://pytorch.org/docs/stable/generated/torch.flatten.html)함
dim=0은 batch_size고 batch_size 이후의 feature(W * H) * kernel 그냥 flatten 하는듯(어차피 이후에 non-linear transform하고
```
import torch.nn as nn
import torchvision.models as models
import torch


class Flatten(nn.Module):
    def __init__(self, dim=-1):
        super(Flatten, self).__init__()
        self.dim = dim

    def forward(self, feat):
        return torch.flatten(feat, start_dim=self.dim)


class ResNetEncoder(models.resnet.ResNet):
    """Wrapper for TorchVison ResNet Model
    This was needed to remove the final FC Layer from the ResNet Model"""
    def __init__(self, block, layers, cifar_head=False, hparams=None):
        super().__init__(block, layers)
        self.cifar_head = cifar_head
        if cifar_head:
            self.conv1 = nn.Conv2d(3 64, kernel_size=3, stride=1, padding=1, bias=False)
            self.bn1 = self._norm_layer(64)
            self.relu = nn.ReLU(inplace=True)
        self.hparams = hparams

        print('** Using avgpool **')

    def forward(self, x):
        x = self.conv1(x)
        x = self.bn1(x)
        x = self.relu(x)
        if not self.cifar_head:
            x = self.maxpool(x)

        x = self.layer1(x)
        x = self.layer2(x)
        x = self.layer3(x)
        x = self.layer4(x)

        x = self.avgpool(x)
        x = torch.flatten(x, 1)

        return x

class ResNet18(ResNetEncoder):
    def __init__(self, cifar_head=True):
        super().__init__(models.resnet.BasicBlock, [2, 2, 2, 2], cifar_head=cifar_head)


class ResNet50(ResNetEncoder):
    def __init__(self, cifar_head=True, hparams=None):
        super().__init__(models.resnet.Bottleneck, [3, 4, 6, 3], cifar_head=cifar_head, hparams=hparams)
```

**Comment by long8v** [comment]
## utils
### LARS.py
```
class LARS(object):
    """
    Slight modification of LARC optimizer from https://github.com/NVIDIA/apex/blob/d74fda260c403f775817470d87f810f816f3d615/apex/parallel/LARC.py
    Matches one from SimCLR implementation https://github.com/google-research/simclr/blob/master/lars_optimizer.py
    Args:
        optimizer: Pytorch optimizer to wrap and modify learning rate for.
        trust_coefficient: Trust coefficient for calculating the adaptive lr. See https://arxiv.org/abs/1708.03888
    """

    def __init__(self,
                 optimizer,
                 trust_coefficient=0.001,
                 ):
        self.param_groups = optimizer.param_groups
        self.optim = optimizer
        self.trust_coefficient = trust_coefficient

    def __getstate__(self):
        return self.optim.__getstate__()

    def __setstate__(self, state):
        self.optim.__setstate__(state)

    def __repr__(self):
        return self.optim.__repr__()

    def state_dict(self):
        return self.optim.state_dict()

    def load_state_dict(self, state_dict):
        self.optim.load_state_dict(state_dict)

    def zero_grad(self):
        self.optim.zero_grad()

    def add_param_group(self, param_group):
        self.optim.add_param_group(param_group)

    def step(self):
        with torch.no_grad():
            weight_decays = []
            for group in self.optim.param_groups:
                # absorb weight decay control from optimizer
                weight_decay = group['weight_decay'] if 'weight_decay' in group else 0
                weight_decays.append(weight_decay)
                group['weight_decay'] = 0
                for p in group['params']:
                    if p.grad is None:
                        continue

                    if weight_decay != 0:
                        p.grad.data += weight_decay * p.data

                    param_norm = torch.norm(p.data)
                    grad_norm = torch.norm(p.grad.data)
                    adaptive_lr = 1.

                    if param_norm != 0 and grad_norm != 0 and group['layer_adaptation']:
                        adaptive_lr = self.trust_coefficient * param_norm / grad_norm

                    p.grad.data *= adaptive_lr

        self.optim.step()
        # return weight decay control to optimizer
        for i, group in enumerate(self.optim.param_groups):
            group['weight_decay'] = weight_decays[i]
```
